### PR TITLE
Read command

### PR DESCRIPTION
**Pull request by dominic31**
Initial support for read and read!.

[Vim documentation, read](http://vimdoc.sourceforge.net/htmldoc/insert.html#:read)
[Vim documentation, read!](http://vimdoc.sourceforge.net/htmldoc/insert.html#:read!)

Range and opt arguments are not yet implemented.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -233,6 +233,7 @@
         "postinstall": "node ./node_modules/vscode/bin/install && gulp init"
     },
     "dependencies": {
+		"child-process": "^1.0.2",
```
Comment (rebornix) [inline package.json]: Indent :)

File: test/cmd_line/read.test.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  setup(async () => {
+    await setupWorkspace();
+    modeHandler = new ModeHandler();
+  });
+
+  teardown(cleanUpWorkspace);
+
+  test("Can read shell command output", async () => {
+    await runCmdLine('r! echo "hey ho"', modeHandler);
```
Comment (rebornix) [inline test/cmd_line/read.test.ts]: I suppose the result of this command differs on different platform, which currently breaks the test on Windows.

**Comment by rebornix** [comment]
Thanks for taking things that are currently not on our list :) Just drop some comments and can you update the roadmap in your PR as well, especially add `Note` to it if `child-process` is not OS-compatible (I'm not sure).

File: test/cmd_line/read.test.ts (hunk 2 above)
Comment (dominic31) [inline test/cmd_line/read.test.ts]: I just didn't realise that Windows prints the "s in echo "..". I modified the test so that it passes on Windows too.

**Comment by dominic31** [comment]
I've updated the roadmap and dropped one comment at least :)

File: src/cmd_line/commands/read.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      return await this.getTextToInsertFromFile();
+    } else if (this.arguments.cmd && this.arguments.cmd.length > 0) {
+      return await this.getTextToInsertFromCmd();
+    } else {
+      throw Error('Invalid arguments');
+    }
+  }
+
+  async getTextToInsertFromFile() : Promise<string> {
+    // TODO: Substitute utf8 with current file encoding, couldn't find this anywhere in vscode's api.
```
Comment (jpoon) [inline src/cmd_line/commands/read.ts]: Is there an issue tracking this with VSCode? Some initial Google says there's some config at the workspace level where you can set the encoding. Maybe we can read that config?

File: test/cmd_line/read.test.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  setup(async () => {
+    await setupWorkspace();
+    modeHandler = new ModeHandler();
+  });
+
+  teardown(cleanUpWorkspace);
+
+  test("Can read shell command output", async () => {
+    await runCmdLine('r! echo heyho', modeHandler);
```
Comment (jpoon) [inline test/cmd_line/read.test.ts]: What happens with a more complex example? 

`:r! echo hey\\n\\tho`

**Comment by jpoon** [comment]
@dominic31 awesome work! dropped a few comments.

File: test/cmd_line/read.test.ts (hunk 4 above)
Comment (dominic31) [inline test/cmd_line/read.test.ts]: 1. This works for me on Ubuntu, probably fails on Windows though :)
2. The error that is thrown is caught in `runCmdLine()` that attempts to display an error message the status bar, this is the same mechanism that would handle  `:thisexcommanddoesnotexist`. I don't see anything in the status bar though, I guess that's another issue.

**Comment by dominic31** [comment]
@jpoon thanks for the comments! I've left some replies.

File: test/cmd_line/read.test.ts (hunk 4 above)
Comment (jpoon) [inline test/cmd_line/read.test.ts]: 1. Can you replace the existing test with `:r! echo hey\\n\\tho`? Helps test more cases :)
2. Yeah, that's a backlog work item to display it on the status bar

**Comment by jpoon** [comment]
LGTM.

* If you choose not to include the detection of the current file encoding, can you file a separate issue so we can track it?
* Update the test

File: test/cmd_line/read.test.ts (hunk 4 above)
Comment (dominic31) [inline test/cmd_line/read.test.ts]: 1. Ah, that's what you meant. Done.

**Comment by dominic31** [comment]
I've updated the unit test. 

About the encoding, there's an [open issue](https://github.com/Microsoft/vscode/issues/824) on the VS Code project for supporting encoding related functions through the api. It hasn't got past the backlog yet. Shall I open an issue here that refers to that issue?

**Comment by jpoon** [comment]
Yes please!

One last thing, can you merge with master and resolve the conflicts?

**Comment by dominic31** [comment]
Sorry, I've gotten things a little muddled up here, the encoding parameter in fs.readFile() is of course the encoding of the external file that gets read in, not the current file in VS Code. Looking at the vim docs for read (link in the first comment), the encoding can be specified in the opt arguments that I haven't yet implemented. I'll update the comment in getTextToInsertFromFile().

Should I open an issue for implementing opt args, or just mention it in the roadmap?

**Comment by johnfn** [comment]
I'm okay with this not having ++opt, that can be a separate task. :)

**Pull request merged**